File: ActivityTuto/app/src/main/java/com/example/activitytuto/MainActivity.java
package com.example.activitytuto;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    Button button1, button2;
    TextView text;
    int count = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.i("LifeCycle", "onCreate() 호출");
        text = findViewById(R.id.text);
        button1 = findViewById(R.id.button1);
        button1.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                count++;
                text.setText("현재 개수=" + count);
            }
        });
        button2 = findViewById(R.id.button2);
        button2.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
            count--;
            text.setText("현재 개수=" + count);
        }
        });

        if (savedInstanceState != null) {
            count = savedInstanceState.getInt("count");
            text.setText("현재 개수=" + count);
        }
    }
    @Override
    public void onStart() {
        super.onStart();
        Log.i("LifeCycle", "onStart() 호출");
    }
    @Override
    public void onResume() {
        super.onResume();
        Log.i("LifeCycle", "onResume() 호출");
    }
    @Override
    public void onPause() {
        super.onPause();
        Log.i("LifeCycle", "onPause() 호출");
    }
    @Override
    public void onStop() {
        super.onStop();
        Log.i("LifeCycle", "onStop() 호출");
    }
    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.i("LifeCycle", "onDestroy() 호출");
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("count", count);
        Log.i("LifeCycle", "onSaveInstanceState() 호출");
    }
}

File: BinderServiceTuto/app/src/main/java/com/example/binderservicetuto/MainActivity.kt
package com.example.binderservicetuto

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.binderservicetuto.LocalService.LocalBinder

class MainActivity : AppCompatActivity() {
    var mService: LocalService? = null
    var mBound = false
    val TAG = "activity"
    private val connection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            val binder = iBinder as LocalBinder
            mService = binder.service
            mBound = true
            Log.i(TAG, "service connect")
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            mBound = false
            Log.i(TAG, "service disConnect")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.i(TAG, "onCreate")
//        val intent = Intent(this, LocalService::class.java)
//        bindService(intent, mConnection, BIND_AUTO_CREATE)
    }

    override fun onStart() {
        super.onStart()
        Intent(this, LocalService::class.java).also {
            bindService(it, connection, Context.BIND_AUTO_CREATE)
        }

        Log.i(TAG, "onStart")
    }
    override fun onResume() {
        super.onResume()
        Log.i(TAG, "onResume")
    }
    override fun onPause() {
        super.onPause()
        Log.i(TAG, "onPause")
    }

    override fun onRestart() {
        super.onRestart()
        Log.i(TAG, "onRestart")
    }
    override fun onStop() {
        super.onStop()
        unbindService(connection)
        mBound = false
        Log.i(TAG, "onStop")
    }
    override fun onDestroy() {
        super.onDestroy()

        Log.i(TAG, "onDestroy")
    }
    //버튼 클릭 이벤트
    fun onButtonClick(v: View?) {
        if (mBound) {
            val num = mService!!.randomNumber
            Toast.makeText(this, "number: $num", Toast.LENGTH_LONG).show()
        }
    }
}